Frequent draw state changes spend time in CanvasStyle variant visitation

kkinnunen-apple opened the pull request:
#### 66dbfd4d25e1ef438fc63ee4eb0662da7f1f0650
<pre>
Frequent draw state changes spend time in CanvasStyle variant visitation
<a href="https://bugs.webkit.org/show_bug.cgi?id=291531">https://bugs.webkit.org/show_bug.cgi?id=291531</a>
<a href="https://rdar.apple.com/149221257">rdar://149221257</a>

Reviewed by Simon Fraser.

During `context.strokeStyle = &quot;rgba(2,2,2,2)&quot;;` WebCore::Color would
be embedded and unembedded in std::variant that CanvasStyle holds.
This would cause noticeable perf loss when done frequently.

Once the code knows the type of the stroke style assignment object,
it should never be held in a variant until stored for some real
polymorphic purpose. Just call normal functions accepting Color instead.

The CanvasStyle is largely a redundant class, remove much of its use.

Fix a bug where canvas string would be stored in state
unparsedStrokeColor/unparsedFillColor even though the color was
unparseable and as such the string and the actual stroke/fill state
color value was the previously set value.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::parseColorWithOverrideAlpha):
(WebCore::CanvasRenderingContext2DBase::setStrokeColorImpl):
(WebCore::CanvasRenderingContext2DBase::setFillColorImpl):
(WebCore::CanvasRenderingContext2DBase::setStrokeColor):
(WebCore::CanvasRenderingContext2DBase::setFillColor):
(WebCore::CanvasRenderingContext2DBase::setStrokeStyle):
(WebCore::CanvasRenderingContext2DBase::setFillStyle):
(WebCore::CanvasRenderingContext2DBase::drawTextUnchecked):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::setStrokeColorImpl):
(WebCore::CanvasRenderingContext2DBase::setFillColorImpl):
* Source/WebCore/html/canvas/CanvasStyle.cpp:
(WebCore::CanvasStyle::CanvasStyle):
(WebCore::CanvasStyle::createFromString): Deleted.
(WebCore::CanvasStyle::createFromStringWithOverrideAlpha): Deleted.
(WebCore::CanvasStyle::isEquivalentColor const): Deleted.
(WebCore::CanvasStyle::applyStrokeColor const): Deleted.
(WebCore::CanvasStyle::applyFillColor const): Deleted.
* Source/WebCore/html/canvas/CanvasStyle.h:
(WebCore::CanvasStyle::CanvasStyle):
(WebCore::CanvasStyle::colorString const):
(WebCore::CanvasStyle::color const):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildInitialState):

* LayoutTests/fast/canvas/canvas-context-save-limit-expected.txt:
Removes one warning message from the expectation.
Before,
`CanvasRenderingContext2DBase::setFillColor(String&amp;&amp; colorString, std::optional&lt;float&gt; alpha)`
would first call `realizeSaves()` that added the warning.
Then it would forward the call to
`CanvasRenderingContext2DBase::setFillStyle(CanvasStyle style)`
which would also call `realizeSaves()` redundantly and add another
warning.

Canonical link: <a href="https://commits.webkit.org/293782@main">https://commits.webkit.org/293782@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cb4a19aa391fbeb974ecce3b5d803983de27ff87

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99945 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105072 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50526 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101986 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19898 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28029 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76088 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33171 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102952 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15185 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90259 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56446 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14990 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8246 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49895 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84914 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8333 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107433 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27058 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19775 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85041 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27421 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86459 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84565 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21471 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29237 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6959 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20881 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26995 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32223 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26806 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30122 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28365 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -564,53 +572,30 @@ void CanvasRenderingContext2DBase::endLayer()
-    checkOrigin(style.canvasPattern().get());
-
     realizeSaves();
+    if (GraphicsContext* c = effectiveDrawingContext())
```
`auto*`

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -564,53 +572,30 @@ void CanvasRenderingContext2DBase::endLayer()
-    checkOrigin(style.canvasPattern().get());
-
     realizeSaves();
+    if (GraphicsContext* c = effectiveDrawingContext())
```
`auto*`

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2113,11 +2089,34 @@ CanvasRenderingContext2DBase::StyleVariant CanvasRenderingContext2DBase::strokeS
+        setStrokeColorImpl(WTFMove(color), WTFMove(colorString));
+        return;
+    }
+    if (std::holds_alternative<RefPtr<CanvasGradient>>(style)) {
```
I'd prefer a blank line after these clauses.

smfr commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -2113,11 +2089,34 @@ CanvasRenderingContext2DBase::StyleVariant CanvasRenderingContext2DBase::strokeS
+        realizeSaves();
+        if (GraphicsContext* c = effectiveDrawingContext())
+            c->setStrokeGradient(gradient->gradient());
+        State& state = modifiableState();
```
`auto&`

kkinnunen-apple committed:
Frequent draw state changes spend time in CanvasStyle variant visitation
https://bugs.webkit.org/show_bug.cgi?id=291531
rdar://149221257

Reviewed by Simon Fraser.

During `context.strokeStyle = "rgba(2,2,2,2)";` WebCore::Color would
be embedded and unembedded in std::variant that CanvasStyle holds.
This would cause noticeable perf loss when done frequently.

Once the code knows the type of the stroke style assignment object,
it should never be held in a variant until stored for some real
polymorphic purpose. Just call normal functions accepting Color instead.

The CanvasStyle is largely a redundant class, remove much of its use.

Fix a bug where canvas string would be stored in state
unparsedStrokeColor/unparsedFillColor even though the color was
unparseable and as such the string and the actual stroke/fill state
color value was the previously set value.

* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::parseColorWithOverrideAlpha):
(WebCore::CanvasRenderingContext2DBase::setStrokeColorImpl):
(WebCore::CanvasRenderingContext2DBase::setFillColorImpl):
(WebCore::CanvasRenderingContext2DBase::setStrokeColor):
(WebCore::CanvasRenderingContext2DBase::setFillColor):
(WebCore::CanvasRenderingContext2DBase::setStrokeStyle):
(WebCore::CanvasRenderingContext2DBase::setFillStyle):
(WebCore::CanvasRenderingContext2DBase::drawTextUnchecked):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
(WebCore::CanvasRenderingContext2DBase::setStrokeColorImpl):
(WebCore::CanvasRenderingContext2DBase::setFillColorImpl):
* Source/WebCore/html/canvas/CanvasStyle.cpp:
(WebCore::CanvasStyle::CanvasStyle):
(WebCore::CanvasStyle::createFromString): Deleted.
(WebCore::CanvasStyle::createFromStringWithOverrideAlpha): Deleted.
(WebCore::CanvasStyle::isEquivalentColor const): Deleted.
(WebCore::CanvasStyle::applyStrokeColor const): Deleted.
(WebCore::CanvasStyle::applyFillColor const): Deleted.
* Source/WebCore/html/canvas/CanvasStyle.h:
(WebCore::CanvasStyle::CanvasStyle):
(WebCore::CanvasStyle::colorString const):
(WebCore::CanvasStyle::color const):
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::buildInitialState):

* LayoutTests/fast/canvas/canvas-context-save-limit-expected.txt:
Removes one warning message from the expectation.
Before,
`CanvasRenderingContext2DBase::setFillColor(String&& colorString, std::optional<float> alpha)`
would first call `realizeSaves()` that added the warning.
Then it would forward the call to
`CanvasRenderingContext2DBase::setFillStyle(CanvasStyle style)`
which would also call `realizeSaves()` redundantly and add another
warning.

Canonical link: https://commits.webkit.org/293782@main
